csharp/src/Google.Protobuf/UnsafeCollectionOperations.cs

```diff
@@ -0,0 +1,118 @@
+    /// a number of significant side effects:
+    /// </para>
+    /// <list type="bullet">
+    /// <item>
+    /// <description>
+    /// The <see cref="Span{T}"/> may or may not reflect the changes made to the
+    /// <see cref="RepeatedField{T}"/>.
+    /// </description>
+    /// </item>
+    /// <item>
+    /// <description>
+    /// If the backing array of the <see cref="RepeatedField{T}"/> is reallocated, further updates
+    /// to the <see cref="Span{T}"/> will not be visible.
+    /// </description>
+    /// </item>
+    /// <item>
+    /// <description>
+    /// The state of the <see cref="Span{T}"/> is undefined after any modifications to the <see
+    /// cref="RepeatedField{T}"/>.
+    /// </description>
+    /// </item>
+    /// </list>
+    /// </remarks>
+    [SecuritySafeCritical]
+    public static class UnsafeCollectionOperations
+    {
+        /// <summary>
+        /// Returns a <see cref="Span{T}"/> that wraps the backing array of the specified
+        /// <see cref="RepeatedField{T}"/>. The returned <see cref="Span{T}"/> should not be used
+        /// after any modifications are made to the <see cref="RepeatedField{T}"/>.
+        /// </summary>
+        /// <typeparam name="T">
+        /// The element type of the <see cref="RepeatedField{T}"/>.
+        /// </typeparam>
+        /// <param name="field">
+        /// The <see cref="RepeatedField{T}"/> to wrap.
+        /// </param>
+        /// <exception cref="ArgumentNullException">
+        /// <paramref name="field"/> is <see langword="null"/>.
+        /// </exception>
+        /// <returns>
+        /// A <see cref="Span{T}"/> that wraps the backing array of the
+        /// <see cref="RepeatedField{T}"/>.
+        /// </returns>
+        public static Span<T> AsSpan<T>(RepeatedField<T> field)
+        {
+            if (field is null)
+            {
+                throw new ArgumentNullException(nameof(field));
+            }
+
+            return field.AsSpan();
+        }
+
+        /// <summary>
+        /// Sets the count of the <see cref="RepeatedField{T}"/> to the specified value.
+        /// This method should only be called if the subsequent code guarantees to populate
+        /// the field with the specified number of items.
+        /// </summary>
+        /// <param name="field">The field to set the count of.</param>
+        /// <param name="count">The value to set the field's count to.</param>
+        /// <typeparam name="T">The type of the elements in the field.</typeparam>
+        /// <exception cref="ArgumentNullException">
+        /// <paramref name="field"/> is <see langword="null"/>.
+        /// </exception>
+        /// <exception cref="ArgumentOutOfRangeException">
+        /// <paramref name="count"/> is negative.
+        /// </exception>
+        public static void SetCount<T>(RepeatedField<T> field, int count)
+        {
+            if (field is null)
```

Comment:
Use `ProtoPreconditions.CheckNotNull(field, nameof(field))` to simplify this and the check above.